Aluno: Fernando Zolubas Preto
Nusp:  10694192

Esse notebook contem apenas a implementacao do algoritimo KNN, o "data preparation" esta no outro notebook.

In [6]:
import numpy as np
import matplotlib.pyplot as plt
import sklearn as skl
import pandas as pd

In [7]:
adult = pd.read_csv(r"C:\Users\Fernando\Desktop\Poli Usp\4 semestre\Machine Learning\adult.csv",
        names=[
        "Age", "Workclass", "fnlwgt", "Education", "Education-Num", "Martial Status",
        "Occupation", "Relationship", "Race", "Sex", "Capital Gain", "Capital Loss",
        "Hours per week", "Country", "Target"],
        sep=r'\s*,\s*',
        engine='python',
        na_values="?")

In [8]:
adult.shape

(32561, 15)

A data base possui 32561 entrevistados; 14 perguntas (atributos); 1 Classificacao (Lable)

In [9]:
adult.head() #printei o "cabecalho dos dados" para visualizar como seria a minha base de dados.

Age         Workclass  fnlwgt  Education  Education-Num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       Martial Status         Occupation   Relationship   Race     Sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   Capital Gain  Capital Loss  Hours per week        Country Target  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

IMPLEMENTANDO O KNN

In [11]:
nadult = adult.dropna() #Retira dados faltantes

In [12]:
nadult.shape

(30162, 15)

In [13]:
#importantdo a base de teste
testAdult = pd.read_csv(r"C:\Users\Fernando\Desktop\Poli Usp\4 semestre\Machine Learning\adult.test.csv",
        names=[
        "Age", "Workclass", "fnlwgt", "Education", "Education-Num", "Martial Status",
        "Occupation", "Relationship", "Race", "Sex", "Capital Gain", "Capital Loss",
        "Hours per week", "Country", "Target"],
        sep=r'\s*,\s*',
        engine='python',
        na_values="?")

In [14]:
nTestAdult = testAdult.dropna()

In [15]:
testAdult.shape

(16282, 15)

In [16]:
nTestAdult.shape

(15060, 15)

Funcao KNN

In [17]:
#Funcao KNN predict
def knn_predict(KN):
    from sklearn.neighbors import KNeighborsClassifier
    from sklearn.model_selection import cross_val_score
    from sklearn.metrics import accuracy_score
    from sklearn import preprocessing
    numAdult = nadult.apply(preprocessing.LabelEncoder().fit_transform)
    numTestAdult = nTestAdult.apply(preprocessing.LabelEncoder().fit_transform)
    Xadult = numAdult.iloc[:,0:14]
    Yadult = numAdult.Target
    XtestAdult = numTestAdult.iloc[:,0:14]
    YtestAdult = numTestAdult.Target
    knn = KNeighborsClassifier(n_neighbors=KN)
    knn.fit(Xadult,Yadult)
    YtestPred = knn.predict(XtestAdult)
    accuracy = accuracy_score(YtestAdult,YtestPred)
    return accuracy

In [45]:
#Funcao KNN predict com selecao de atributos
def knn_predict(KN,Atributos):
    from sklearn.neighbors import KNeighborsClassifier
    from sklearn.model_selection import cross_val_score
    from sklearn.metrics import accuracy_score
    from sklearn import preprocessing
    numAdult = nadult.apply(preprocessing.LabelEncoder().fit_transform)
    numTestAdult = nTestAdult.apply(preprocessing.LabelEncoder().fit_transform)
    Xadult = numAdult.iloc[:,Atributos]
    Yadult = numAdult.Target
    XtestAdult = numTestAdult.iloc[:,Atributos]
    YtestAdult = numTestAdult.Target
    knn = KNeighborsClassifier(n_neighbors=KN)
    knn.fit(Xadult,Yadult)
    YtestPred = knn.predict(XtestAdult)
    accuracy = accuracy_score(YtestAdult,YtestPred)
    return accuracy

In [18]:
#Funcao KNN com cross validation
def knn_cv(KN,predictt):
    pcv = 10;
    from sklearn.neighbors import KNeighborsClassifier
    from sklearn.model_selection import cross_val_score
    from sklearn.metrics import accuracy_score
    from sklearn import preprocessing
    numAdult = nadult.apply(preprocessing.LabelEncoder().fit_transform)
    numTestAdult = nTestAdult.apply(preprocessing.LabelEncoder().fit_transform)
    Xadult = numAdult.iloc[:,0:14]
    Yadult = numAdult.Target
    XtestAdult = numTestAdult.iloc[:,0:14]
    YtestAdult = numTestAdult.Target
    knn = KNeighborsClassifier(n_neighbors=KN)
    scores_cv = cross_val_score(knn, Xadult, Yadult, cv=pcv)
    
    i=0
    average = 0
    while i<pcv :
        average = average + scores_cv[i]
        i = i + 1
    average = average/pcv
    
    if(predictt == 1):
        knn.fit(Xadult,Yadult)
        YtestPred = knn.predict(XtestAdult)
        print(YtestPred)
    return average

In [22]:
#Funcao KNN com cross validation
def knn_cv(KN,PrimeiroAtributo,UltimoAtributo):
    pcv = 10;
    from sklearn.neighbors import KNeighborsClassifier
    from sklearn.model_selection import cross_val_score
    from sklearn.metrics import accuracy_score
    from sklearn import preprocessing
    numAdult = nadult.apply(preprocessing.LabelEncoder().fit_transform)
    numTestAdult = nTestAdult.apply(preprocessing.LabelEncoder().fit_transform)
    Xadult = numAdult.iloc[:,PrimeiroAtributo:UltimoAtributo]
    Yadult = numAdult.Target
    XtestAdult = numTestAdult.iloc[:,PrimeiroAtributo:UltimoAtributo]
    YtestAdult = numTestAdult.Target
    knn = KNeighborsClassifier(n_neighbors=KN)
    scores_cv = cross_val_score(knn, Xadult, Yadult, cv=pcv)
    
    i=0
    average = 0
    while i<pcv :
        average = average + scores_cv[i]
        i = i + 1
    average = average/pcv
    
        
    return average

In [26]:
#Funcao KNN com cross validation
def knn_cv_select(KN,Atributos):
    pcv = 10;
    from sklearn.neighbors import KNeighborsClassifier
    from sklearn.model_selection import cross_val_score
    from sklearn.metrics import accuracy_score
    from sklearn import preprocessing
    numAdult = nadult.apply(preprocessing.LabelEncoder().fit_transform)
    numTestAdult = nTestAdult.apply(preprocessing.LabelEncoder().fit_transform)
    Xadult = numAdult.iloc[:,Atributos]
    Yadult = numAdult.Target
    XtestAdult = numTestAdult.iloc[:,Atributos]
    YtestAdult = numTestAdult.Target
    knn = KNeighborsClassifier(n_neighbors=KN)
    scores_cv = cross_val_score(knn, Xadult, Yadult, cv=pcv)
    
    i=0
    average = 0
    while i<pcv :
        average = average + scores_cv[i]
        i = i + 1
    average = average/pcv
    
        
    return average

Testes iniciais, procurando uma faixa de pesquisa adequada para se encontrar o numero de vizinhos que otimiza a acuracia do KNN

In [14]:
knn_cv(1,1)

[0 0 0 ... 0 1 0]


0.7279025607708232

In [24]:
knn_cv(10,0)

0.7839673784403363

In [25]:
knn_cv(100,0)

0.7513096103325829

In [26]:
knn_cv(500,0)

0.7510775258171767

In [27]:
knn_cv(1000,0)

0.7510775258171767

In [28]:
knn_cv(10000,0)

0.7510775258171767

In [15]:
knn_cv(27000,0)

0.7510775258171767

Vemos que a partir de K=500 nao obtemos melhora no resultado do KNN. Vamos olhar mais alguns casos menores do que 500 

In [32]:
knn_cv(200,0)

0.7510775258171767

In [30]:
knn_cv(300,0)

0.7510775258171767

In [31]:
knn_cv(400,0)

0.7510775258171767

In [33]:
knn_cv(150,0)

0.7510775258171767

In [15]:
knn_cv(125,0)

0.7511106933130308

Nao houve uma resposta adequada a tentativa automatica de procura de otimizacao dos hiperparametro K. Sendo assim decidiu-se realizar tal processo manualmente
i=1
max_cv_score = 0
while i<100:
    if (knn_cv(i,0) > max_cv_score):
        max_cv_score = knn_cv(i,0)
        i = i + 40
print(max_cv_score)
print(i)

In [13]:
knn_cv(20,0)

0.7862212400606577

In [14]:
knn_cv(30,0)

0.7785626269633946

In [15]:
knn_cv(40,0)

0.7686492839715845

In [16]:
knn_cv(50,0)

0.7612560988404669

In [17]:
knn_cv(60,0)

0.7567138121473423

In [18]:
knn_cv(70,0)

0.7539620314486931

Procura de 10 a 30

In [19]:
knn_cv(15,0)

0.7855584838163504

In [20]:
knn_cv(25,0)

0.7839668947831064

Procura de 15 a 30. Temos 15,20,25,30

In [21]:
knn_cv(17,0)

0.7869839385412274

In [22]:
knn_cv(23,0)

0.7849284333321267

In [23]:
knn_cv(29,0)

0.7808503046288934

Procura de 16 ate 19

In [24]:
knn_cv(16,0)

0.7877134584923119

In [25]:
knn_cv(18,0)

0.7877795955930952

In [26]:
knn_cv(19,0)

0.7875807445274645

O cross validation aponta que o algoritimo responde melhor, dado os atributos selecionados, quando k = 18. Facamos o teste na base de testes

In [27]:
knn_predict(18)

0.7903054448871182

Apenas otimizando o numero de vizinhos do algoritimo knn sem levar em conta o data prep onde se escolhe os atribibutos que serao levados em conta obteve-se uma acuracia de 79,03%

Usando agora as analises do data prep sera feita uma tentativa de otimizacao do resultado obtido anteriormente em que se encontrou o melhor numero de vizinhos quando-se considera todos os atributos.

Na etapa seguinte serao selecionados os vizinhos de interesse.

In [ ]:
Testando a funcao knn_cv_select

In [23]:
knn_cv(16,0,14)

0.7877134584923119

In [24]:
knn_cv(16,0,10)

0.7459388726896814

In [27]:
knn_cv_select(16,[0,1,2,3,4,5,6,7,8,9])

0.7459388726896814

O resultado do data prep apontou para a selecao dos seguintes atributos:
Age 0
Education 3
Occupation 6
Relationship 7
Capital gain 10
Capital loss 11

Onde os numeros sao as indexacoes desses atributos na funcao knn_cv_select que permite selecionar os atributos selecionados na analise do knn retornado a media do cross validation.

In [28]:
knn_cv_select(16,[0,3,6,7,10,11])

0.8344279227975984

É possivel perceber que a selecao de atributos atarves do processo de data prep, quando basedado nas indicacoes assinaladas na literatura de machine learning bem como em analises dos dados tomando-se decisoes simplismente pelo bom senso do que se observa nas analises resulta em grande melhoria na qualidade do classificador. 

Para essa versao final do classificador sera feita, a seguir, a procura do numero de vizinhos que otimiza o classificador de parametros (knn_otimizador,[0,3,6,7,10,11]) considerando-se esse o melhor classificador dado o resultado obtido na linha superior e os limites de tempo disponivel para a realizacao dessa atividade.

In [31]:
knn_cv_select(1,[0,3,6,7,10,11])

0.8006100070714822

In [32]:
knn_cv_select(10,[0,3,6,7,10,11])

0.8358868748462573

In [33]:
knn_cv_select(20,[0,3,6,7,10,11])

0.8336983259391528

In [34]:
knn_cv_select(30,[0,3,6,7,10,11])

0.833201055432019

In [35]:
knn_cv_select(40,[0,3,6,7,10,11])

0.8316096423026291

In [36]:
knn_cv_select(5,[0,3,6,7,10,11])

0.82975387865374

In [37]:
knn_cv_select(7,[0,3,6,7,10,11])

0.835456170162266

In [38]:
knn_cv_select(8,[0,3,6,7,10,11])

0.8356550322032067

In [39]:
knn_cv_select(9,[0,3,6,7,10,11])

0.8332016819724014

In [41]:
knn_cv_select(11,[0,3,6,7,10,11])

0.8359200642927311

In [42]:
knn_cv_select(12,[0,3,6,7,10,11])

0.8355221093429641

In [43]:
knn_cv_select(13,[0,3,6,7,10,11])

0.8354559172781496

In [44]:
knn_cv_select(14,[0,3,6,7,10,11])

0.8360194348724506

In [46]:
knn_predict(11,[0,3,6,7,10,11])

0.8258300132802124

Conclusão:

Realizou-se o data prep em que se observou as distribuicoes das subclasses de cada atributo procurando-se entender melhor suas caracteristicas por meio da funcao: ganhaPouco()

Comparou-se o classificador nas situacoes sem e com as consideracoes do data prep em que se obteve-se uma acuracia de 79,03% e 82,58% respectivamente.

Assim, dada a limitacao de tempo e de processamento computacional chegou-se a conclusao nesse estudo de que uma boa opcao de classificador KNN para a base de dados adult seria aquele que leva em consideração os atributos "age", "Education", "Ocuppation", "Relationship", "Capital gain" e "capital loss" considerando-se 11 vizinhos se o objetivo for classificicar a renda anual de uma pessoa nas condicoes dessa base de dados.